lib/THCUNN/SparseLinear.cu

```diff
@@ -0,0 +1,243 @@
+#include "THCUNN.h"
+#include <stdio.h>
+#include <cusparse.h>
+
+#if CUDA_VERSION >= 7050
+static cusparseHandle_t cusparse_handle = 0;
+
+static void init_cusparse() {
+  if (cusparse_handle == 0) {
+    cusparseStatus_t status = cusparseCreate(&cusparse_handle);
+    if (status != CUSPARSE_STATUS_SUCCESS) {
+      THError("CUSPARSE Library initialization failed");
+    }
+  }
+}
+#endif
+
+static bool checkInput(THCudaTensor* t)
+{
+  return t->nDimension == 2 && t->size[1] == 2;
+}
+
+static bool checkSize2D(THCudaTensor* t, long size0, long size1)
+{
+  return t->nDimension == 2 && t->size[0] == size0 && t->size[1] == size1;
+}
+
+static bool checkSize1D(THCudaTensor* t, long size0)
+{
+  return t->nDimension == 1 && t->size[0] == size0;
+}
+
+void THNN_CudaSparseLinear_updateOutput(THCState *state, THCudaTensor *input,
+          THCudaTensor *output, THCudaTensor *weight, THCudaTensor *bias,
+          THCudaTensor *shardBuffer)
+{
+  long outDim = weight->size[0];
+  long inDim = weight->size[1];
+  long nnz = input->size[0];
+
+  THArgCheck(checkInput(input), 2, "input size must be nnz x 2");
+  THArgCheck(checkSize1D(output, outDim), 3, "output size wrong");
+  THArgCheck(checkSize1D(bias, outDim), 5, "bias size wrong");
+
+#if CUDA_VERSION < 7050
+  THCudaTensor_copy(state, output, bias);
+  for (int i = 0; i < nnz; i++)
+  {
+    long offset = (long)(THCudaTensor_get2d(state, input, i, 0)) - 1;
+    if (offset >= 0 && offset < inDim) // make sure indices are in bounds..
+    {
+      float val = THCudaTensor_get2d(state, input, i, 1);
+      THCudaBlas_axpy(state,
+          output->size[0],
+          val,
+          THCudaTensor_data(state, weight)+offset*weight->stride[1],
+          weight->stride[0],
+          THCudaTensor_data(state, output),
+          output->stride[0]
+          );
+    }
+    else
+    {
+      THError("index out of bound. updateOutput: \
+          %ld not between 1 and %ld", offset + 1, inDim);
+    }
+  }
+#else
+  void *pBuffer;
+  int pBufferSize = 0;
+  // output = bias
+  THCudaTensor_copy(state, output, bias);
+
+  // Get data ready for cusparse
+  THCudaTensor *values = THCudaTensor_newContiguous(state, THCudaTensor_newSelect(state, input, 1, 1));
+  THCudaTensor *indicies = THCudaTensor_newSelect(state, input, 1, 0);
+  THCudaIntTensor *indicies_int = THCudaIntTensor_newWithSize1d(state, nnz);
+  THCudaIntTensor_copyCudaFloat(state, indicies_int, indicies);
+
+  init_cusparse();
+  float one = 1;
+  cusparseSgemvi_bufferSize(cusparse_handle,
+          CUSPARSE_OPERATION_TRANSPOSE,
+          inDim,
+          outDim,
+          nnz,
+          &pBufferSize
+  );
+  cudaMalloc((void**)&pBuffer, pBufferSize);
```

Comment:
I think since this is part of an nn layer, you can initialize it in nn, keep it around and pass it in. You can call THCudaTensor_resize() which will just resize if it needs a bigger buffer.
A similar line of code would be:
https://github.com/torch/nn/blob/master/SpatialConvolution.lua#L51
https://github.com/torch/nn/blob/master/SpatialConvolution.lua#L109
https://github.com/torch/nn/blob/master/SpatialConvolution.lua#L180

and follow the "columns" variable in here:
https://github.com/torch/cunn/blob/master/lib/THCUNN/SpatialConvolutionMM.cu